smart-contract-extensions/contracts/KeyPurchaser.sol

```diff
@@ -0,0 +1,162 @@
+pragma solidity 0.5.16;
+
+import 'hardlydifficult-ethereum-contracts/contracts/lifecycle/Stoppable.sol';
+import '@openzeppelin/upgrades/contracts/Initializable.sol';
+import '@openzeppelin/contracts/utils/Address.sol';
+import '@openzeppelin/contracts/token/ERC20/IERC20.sol';
+import '@openzeppelin/contracts/token/ERC20/SafeERC20.sol';
+import 'unlock-abi-1-3/IPublicLockV6.sol';
+
+/**
+ * @notice Purchase a key priced in any ERC-20 token - either once or as a regular subscription.
+ * This allows the user to purchase or subscribe to a key with 1 tx (`approve`)
+ * or if the token supports it, with 1 signed message (`permit`).
+ *
+ * The user can remove approval to cancel anytime.
+ *
+ * Risk: if the user transfers or cancels the key, they would naturally expect that also cancels
+ * the subscription but it does not. This should be handled by the frontend.
+ */
+contract KeyPurchaser is Initializable, Stoppable
+{
+  using Address for address payable;
+  using SafeERC20 for IERC20;
+
+  // set on initialize and cannot change
+  IPublicLock public lock;
+  uint public maxKeyPrice;
+  uint public maxPurchaseCount;
+  uint public renewWindow;
+  uint public renewMinFrequency;
+  bool public isSubscription;
+
+  // admin can change these anytime
+  string public name;
+  bool internal disabled;
+
+  // store minimal history
+  mapping(address => uint) public timeOfLastPurchase;
+
+  /**
+   * @notice Called once to set terms that cannot change later on.
+   */
+  function initialize(
+    IPublicLock _lock,
+    uint _maxKeyPrice,
+    uint _renewWindow,
+    uint _renewMinFrequency,
```

Comment:
If you could add some @param comments above it will probably save me a bunch of questions further down in purchaseFor()

Reply:
done